src/selectors/entities/channels.js

```diff
@@ -754,3 +823,172 @@ export const filterPostIds = (condition) => {
         }
     );
 };
+
+const getAllActiveChannels = createSelector(
+    getPublicChannels,
+    getPrivateChannels,
+    getDirectChannels,
+    (
+        publicChannels,
+        privateChannels,
+        directChannels,
+    ) => {
+        const allChannels = [
+            ...publicChannels,
+            ...privateChannels,
+            ...directChannels,
+        ];
+
+        return allChannels;
+    }
+);
+
+export const getAllChannelIds = createIdsSelector(
+    getAllActiveChannels,
+    getCurrentUser,
+    getMyChannelMemberships,
+    getLastPostPerChannel,
+    (state, lastUnreadChannel, unreadsAtTop, favoritesAtTop, sorting = 'alpha') => sorting,
+    mapAndSortChannelIds,
+);
+
+export const getAllSortedChannelIds = createIdsSelector(
+    getUnreadChannelIds,
+    getFavoritesPreferences,
+    getAllChannelIds,
+    (state, lastUnreadChannel, unreadsAtTop = true) => unreadsAtTop,
+    (state, lastUnreadChannel, unreadsAtTop, favoritesAtTop = true) => favoritesAtTop,
+    filterChannels,
+);
+
+export const getOrderedChannelIds = (state, lastUnreadChannel, grouping, sorting, unreadsAtTop, favoritesAtTop) => {
```

Comment:
This needs to be memoized since the Sidebar component in the webapp and the ChannelsList in the mobile app are calling this in mapStateToProps, but that's understandably hard because of how much goes into this function to make the output possibly changed.

While `createSelector` won't work because this requires the whole state be passed in, you could attempt to memoize it after the sections have been created since `getSortedPublicChannelIds` and such appear to be memoized correctly. Something like this would probably work to return the same channel arrays even if something unrelated in the store has changed.
```javascript
export const getOrderedChannelIds = () => {
    let lastChannels;

    const hasChannelsChanged = (channels) => {
        if (!lastChannels || lastChannels.length !== channels.length) {
            return true;
        }

        for (let i = 0; i < channels.length; i++) {
            if (channels[i].type !== lastChannels[i].type || channels[i].items !== lastChannels[i].items) {
                return true;
            }
        }

        return false;
    };

    return (state, lastUnreadChannel, grouping, sorting, unreadsAtTop, favoritesAtTop) => {
        const channels = ...; // Do everything like it's currently done to construct channels

        if (hasChannelsChanged(channels)) {
            lastChannels = channels;
        }

        return lastChannels;
    };
}();
```